pkg/cmd/roachtest/roachtestutil/mixedversion/mutators.go

```diff
@@ -468,19 +488,194 @@ func (m panicNodeMutator) Generate(
+			Filter(func(s *singleStep) bool {
+				// We don't want to set up a partition concurrently with other steps, and inserting
+				// before a concurrent step causes the step to run concurrently with that step, so
+				// we filter out any concurrent steps. We don't want to set up a partition while
+				// nodes are unavailable, as that could cause the cluster to lose quorum,
+				//	so we filter out steps with unavailable nodes.
+				var unavailableNodes bool
+				if planner.isMultitenant() {
+					unavailableNodes = s.context.Tenant.hasUnavailableNodes || s.context.System.hasUnavailableNodes
+				} else {
+					unavailableNodes = s.context.System.hasUnavailableNodes
+				}
+				return s.context.System.Stage >= InitUpgradeStage && !idx.IsConcurrent(s) && !unavailableNodes
+			})
+
+		stepToPartition := possiblePointsInTime.RandomStep(rng)
+		hasInvalidConcurrentStep := false
+		var firstStepInConcurrentBlock *singleStep
+
+		isInvalidRecoverStep := func(s *singleStep) bool {
+			// Restarting a node in the middle of a network partition has a chance of
+			// loss of quorum, so we do should recover the network partition before this
+			// if the restarted node is not the node being partitioned.
+			// TODO: The partitioned node should be able to restart safely, provided
+			// the necessary steps are altered to allow it.
+
+			_, restartSystem := s.impl.(restartWithNewBinaryStep)
+			_, restartTenant := s.impl.(restartVirtualClusterStep)
+			// Many hook steps require communication between specific nodes, so we
+			// should recover the network partition before running them.
+			_, runHook := s.impl.(runHookStep)
+
+			if idx.IsConcurrent(s) {
+				if firstStepInConcurrentBlock == nil {
+					firstStepInConcurrentBlock = s
+				}
+				hasInvalidConcurrentStep = true
+			} else {
+				hasInvalidConcurrentStep = false
+				firstStepInConcurrentBlock = nil
+			}
+
+			var unavailableNodes bool
+			if planner.isMultitenant() {
+				unavailableNodes = s.context.Tenant.hasUnavailableNodes || s.context.System.hasUnavailableNodes
+			} else {
+				unavailableNodes = s.context.System.hasUnavailableNodes
+			}
+			return unavailableNodes || restartTenant || restartSystem || runHook
+		}
+
+		_, validStartStep := upgrade.CutAfter(func(s *singleStep) bool {
+			return s == stepToPartition[0]
+		})
+
+		validEndStep, _, cutStep := validStartStep.Cut(func(s *singleStep) bool {
+			return isInvalidRecoverStep(s)
+		})
+
+		// Inserting before a concurrent step will cause the step to run concurrently with that step,
+		// so we remove the concurrent steps from the list of possible insertions if they contain
+		// any invalid steps.
+		if hasInvalidConcurrentStep {
+			validEndStep, _ = validEndStep.CutAfter(func(s *singleStep) bool {
+				return s == firstStepInConcurrentBlock
+			})
+		}
+
+		nodeCount := len(nodeList)
+		partitionedNode := nodeList[rng.Intn(nodeCount)]
+
+		leftPartition := []install.Node{install.Node(partitionedNode)}
```

Comment:
The way we pick these partition groups makes it so we usually test partial partitions instead of total partitions.  I'm not sure what the real world odds are, but my intuition is that it's not a flat % chance for each node to be partitioned and that total partitions are much more likely. 

What do you think about making it a 50% chance we just default to a total partition? i.e. we pick a random node and partition it from the rest of the cluster?

Reply:
Yeah I had the percentage at 70% per node to increase the chances of a total partition, but this makes more sense to me. In that case then, would 50% per node be appropriate in the event that it is a partial partition?